Fix compute_test_value error when creating observed variables

vandalt opened the pull request:
<!-- !! Thank your for opening a PR !! -->

**What is this PR about?**
This fixes #6981. The issue was about observed variables causing a `TypeError` when a model is created with `pytensor.config.compute_test_value != "off"` It adds a small test that reproduces the issue and replaces `rv_var.shape` by `rv_var.type.shape` in the `np.broadcast_to()` call.

Let me know if I should move the test elsewhere or if something else is needed!

Thanks!

**Checklist**
+ [x] Explain important implementation details 👆
+ [x] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [x] Link relevant issues (preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html))
+ [x] Are the changes covered by tests and docstrings?
+ [x] Fill out the short summary sections 👇

## Bugfixes
- Fixes #6981.

<!-- readthedocs-preview pymc start -->
----
:books: Documentation preview :books:: https://pymc--6982.org.readthedocs.build/en/6982/

<!-- readthedocs-preview pymc end -->

vandalt committed:
Add failing test for observed RVs with `compute_test_value='false'`

vandalt committed:
Use `rv_var.type.shape` instead of `rv_var.shape` when testing observed RV in model creation (Fix #6981)

- Issue: https://github.com/pymc-devs/pymc/issues/6981
- This fix only affects cases with `pytensor.config.compute_test_value!="off"`. Code is not executed otherwise.

ricardoV94 commented on pymc/model/core.py:
```diff
@@ -1352,7 +1352,7 @@ def make_obs_var(
-                rv_var.tag.test_value = np.broadcast_to(test_value, rv_var.shape)
+                rv_var.tag.test_value = np.broadcast_to(test_value, rv_var.type.shape)
             else:
                 rv_var.tag.test_value = data
```
There could be None in `type.shape`. Broadcasting shouldn't be needed because the RV should always have the right shape.

I suggest simplifying the logic to:
```
if test_value is None:
    rv_var.tag.test_value = data
```
And ignore the whole broadcast logic

ricardoV94 replied:
I can't think of a scenario where this could be None though because the RV is created by the time this method is called. Can we just remove this section al together?

vandalt replied:
I implemented the first suggestion.

I can't see a case where `test_value` is None either, but you're much more familiar with this code than I am so I'll implement the option you prefer! :) Should I just remove the `if pytensor.config.compute_test_value != "off":` block completely?

ricardoV94 replied:
Yes, let's do that and see if any tests fails

vandalt replied:
Done! I ran tests in `test_core.py` locally, but not the whole suite as I'm guessing it will run here.

ricardoV94 commented on tests/model/test_core.py:
```diff
@@ -223,6 +223,20 @@ def test_observed_type(self):
         assert x2.type.dtype == X.type.dtype
 
+    def test_observed_compute_test_value(self):
+        try:
```
No need for the try except

ricardoV94 commented on tests/model/test_core.py:
```diff
@@ -223,6 +223,20 @@ def test_observed_type(self):
+                rng = np.random.default_rng(123)
+                data = rng.standard_normal(100)
+                with pm.Model():
+                    obs = pm.Normal(
```
should test something about the tag.test_value of obs

vandalt replied:
I added a test that compares `shape` and `dtype` with `data`. Let me know if you had something else in mind!

ricardoV94 reviewed (commented):
Thanks for opening a PR.

This is related to some functionality that is not much needed in PyMC anymore and we may remove this PyTensor flag in the near future. Anyway, I've left some comments below

vandalt committed:
Remove broadcast logic from test_value check

Should always have the right shape when not None

vandalt committed:
Remove unncessary try/except in `test_observed_compute_test_value()`

vandalt committed:
Test that shape and dtype of observed normal distribution are the same as observed data

vandalt committed:
Remove `compute_test_value != "off"` logic completely in make_obs_var

There are no cases where the `test_value` should be None, so the code inside that block should not ever be used.

ricardoV94 committed:
Simplify test
